### PR TITLE
mgr/dashboard: Unify Tasks and Notifications into a sidebar

### PR DESCRIPTION
**Pull request by tspmelo**
Fixes: https://tracker.ceph.com/issues/37402

Signed-off-by: Tiago Melo <tmelo@suse.com>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by tspmelo** [comment]
localhost_4200_ (1)

**Comment by tspmelo** [comment]
I wrote some questions in the [issue](https://tracker.ceph.com/issues/37402#note-4).

**Comment by votdev** [comment]
The commit message has a typo: s/Taks/Task/

**Comment by LenzGr** [comment]
Nit: there's a small typo in the commit message: s/Taks/Tasks/

**Comment by tspmelo** [comment]
jenkins retest this please

**Comment by tspmelo** [comment]
jenkins retest this please

**Comment by tspmelo** [comment]
jenkins test make check

**Comment by callithea** [comment]
Tested it locally, I only see processing long running tasks, the notifications e.g. about an updated pool are not displayed in the sidebar.. not sure if someone can reproduce this?

**Comment by tspmelo** [comment]
@callithea Do you mean the notification that a pool is updating or that it finished updating?

**Comment by callithea** [comment]
> @callithea Do you mean the notification that a pool is updating or that it finished updating?

@tspmelo I can only see that it's updating a pool when the update is still ongoing. But when the update is done I can't see any notifications that I upated the pool. Also I can't see something like you posted in the screenshot about "Created pool xy"  when I created a pool.

**Comment by tspmelo** [comment]
Just tested the PR and got this:
notification

Are you getting any errors in the browser console?

**Comment by callithea** [comment]
Small update: I can see notifications for e.g. OSDs, but still don't see any when creating/editing pools/RBDs, need to investigate..


Screenshot_20190828_223937

**Comment by tspmelo** [comment]
jenkins test dashboard

**Comment by p-se** [comment]
> Small update: I can see notifications for e.g. OSDs, but still don't see any when creating/editing pools/RBDs, need to investigate..

I can reproduce that issue. I see popup notifications when creating and deleting an OSD, but don't see any information about that in the sidebar.

Is that on purpose?

> I can also see the pop up messages for all actions when it comes to creating/editing/deleting items. But they do not appear in the sidebar notifcation list (except for OSD notifications).
> 
> [screenshot removed]

**Comment by tspmelo** [comment]
@callithea @p-se Should be fixed now.
I was not able to reproduce it because my local storage already had value for one of the vars.

**Comment by callithea** [comment]
> @callithea @p-se Should be fixed now.
> I was not able to reproduce it because my local storage already had value for one of the vars.

@tspmelo Yes, works for me now, thanks! Will approve as soon as the checks passed.

**Comment by callithea** [comment]
jenkins test dashboard

**Comment by tspmelo** [comment]
jenkins test dashboard

**Comment by callithea** [comment]
I was able to reproduce the error of the dashboard tests on my local system several times (based on the master branch):

`` W/element - more than one element found for locator By(css selector, .card-header) - the first result will be used
    fields check
      ✗ should check that title & table appears
        - Expected '' to match 'CRUSH map viewer'.``

Maybe there's a way to address this in #30017 ?

**Comment by tspmelo** [comment]
Found the problem. One of the css expression was catching an element from the sidebar, causing the failure in e2e.

Also had to move some code from the notification service to the notification-sidebar component.

**Comment by tspmelo** [comment]
duration

Which position do you like most?

**Comment by votdev** [comment]
> duration
> 
> Which position do you like most?

I vote for using the ``Duration: xxx seconds`` version because displaying ``(xxx seconds)`` in the title is confusing because it does not mention what this time is for.

**Comment by LenzGr** [comment]
> I vote for using the `Duration: xxx seconds` version because displaying `(xxx seconds)` in the title is confusing because it does not mention what this time is for.

Agreed, that makes it more clear. But then I wonder if we should display both the time stamp *and* the time elapsed since this event occurred.

I think this part should be simplified to read: "Finished on: [time stamp]" or "Finished [duration] ago". Thoughts?

**Comment by tspmelo** [comment]
@LenzGr  the duration is not how long ago it finished, it's how long it took to finish the task.
So the timestamp and duration are showing 2 different information.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/notifications-sidebar/notifications-sidebar.component.html [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+         aria-hidden="true"></i>
+      &nbsp;
+      <ng-container i18n>Clear notifications</ng-container>
+    </button>
+
+    <hr>
+
+    <div *ngFor="let notification of notifications">
+      <div class="card border-0 mb-3"
+           style="max-width: 540px;">
```
[the remark this answers is not in the document]
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/notifications-sidebar/notifications-sidebar.component.html]: I don't really consider setting styles in the style attribute polluting the HTML, but I will change it.

**Comment by ricardoasmarques** [comment]
+1 for `Duration: xxx seconds`.

**Comment by LenzGr** [comment]
>  the duration is not how long ago it finished, it's how long it took to finish the task.
> So the timestamp and duration are showing 2 different information.

That was exactly my point ;) - but you're also putting a duration after the time stamp, that should either be improved by adding the string "ago" behind it, to make clear that these are two different time intervals.

The point I would like to clarify: do we need to print both the time stamp and the time elapsed since that time stamp, or is providing one of them enough?

We could either do **a)**:

[Notification text]
Duration: [duration]
[Time stamp] ([time elapsed] ago)

or **b)**:

[Notification text]
Duration: [duration]
[Time stamp]

or **c)**:

[Notification text]
Duration: [duration]
([time elapsed] ago)

**Comment by tspmelo** [comment]
Current implementation:
Screenshot_20190911_1052331

the title overlay is darker, but for some reason, when I try to take a screenshot it fades...

**Comment by callithea** [comment]
jenkins test dashboard

**Comment by callithea** [comment]
jenkins test docs

**Comment by callithea** [comment]
Is it on purpose that the closing button ("X") in the top right corner is surrounded by a light blue box?
> Screenshot_20190911_1052331

In Firefox it's surrounded by grey dots:

Screenshot_20190918_113502

**Comment by tspmelo** [comment]
@callithea Not on purpose.
Browsers seem to auto focus the button each time the sidebar is opened.
I will try to disable it.

**Comment by callithea** [comment]
> @callithea Not on purpose.
> Browsers seem to auto focus the button each time the sidebar is opened.
> I will try to disable it.

Re-tested it locally, it's disabled now.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/notifications-sidebar/notifications-sidebar.component.html [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+          </div>
+          <div class="col-md-9">
+            <div class="card-body p-0">
+              <h6 class="card-title bold">{{ notification.title }}</h6>
+              <p class="card-text"
+                 [innerHtml]="notification.message"></p>
+              <p class="card-text text-muted">
+                <ng-container *ngIf="notification.duration">
+                  <small>
+                    <ng-container i18n>Duration:</ng-container> {{notification.duration | duration}}
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/notifications-sidebar/notifications-sidebar.component.html]:
```suggestion
                    <ng-container i18n>Duration:</ng-container> {{ notification.duration | duration }}
```

**Comment by votdev** [comment]
A new notification shows ``a few seconds ago`` though the timestamp (using the browser dev tools) shows that the notification appears 5 minutes ago. The text changes after a page reload to ``5 minutes ago``. I used Chromium for testing.

**Comment by tspmelo** [comment]
@votdev Fixed.

**Comment by tspmelo** [comment]
jenkins retest this please

**Pull request merged**